src/debugpy/server/tracing/tracer.py

```diff
@@ -0,0 +1,417 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root
+# for license information.
+
+# Once callbacks are registered they are invoked even during finalization when the
+# Python is shutting down. Thus, trace_* methods, and any other methods that they
+# invoke, must not use any globals from this or other modules (including globals
+# that represent imported modules or defined classes!) until it checks that they
+# are present, or preload them into class or instance attributes in advance.
+# To facilitate this, Tracer is defined in a separate submodule which should not
+# contain ANY top-level imports other than typing nor definitions other than the
+# class itself. All other imports must be done in class scope and then referred to
+# from methods via self.
+
+from types import CodeType, FrameType
+from typing import Iterable
+
+
+class Log:
+    """
+    Safe logging for Tracer. Delegates to debugpy.common.log, but only when it is
+    safe to do so (i.e. not during finalization).
+    """
+
+    from debugpy.common import log
+
+    def __init__(self):
+        import atexit
+
+        def nop(*args, **kwargs):
+            pass
+
+        @atexit.register
+        def disable():
+            self.debug = self.info = self.warning = self.error = self.exception = nop
+
+    def debug(self, *args, **kwargs):
+        # self.log.debug("{0}", *args, **kwargs)
+        # print(*args)
+        pass
+
+    def info(self, *args, **kwargs):
+        self.log.info("{0}", *args, **kwargs)
+
+    def warning(self, *args, **kwargs):
+        self.log.warning("{0}", *args, **kwargs)
+
+    def error(self, *args, **kwargs):
+        self.log.error("{0}", *args, **kwargs)
+
+    def exception(self, *args, **kwargs):
+        self.log.exception("{0}", *args, **kwargs)
+
+
+class Tracer:
+    """
+    Singleton that manages sys.monitoring callbacks for this process.
+    """
```

Comment:
A lot of this code used to be in tracing.py. Is there a reason to put it in a singleton class? I understand the needs for singletons in languages that force everything to be classes, but why in Python?

Note - I'm not saying undo it, just wondering what informed the design here.

Reply:
Two reasons. First one is that experimenting with it in the prototype leads me to believe that writing similar code as global functions is more fraught with footguns in general, thanks to the need to declare all writable shared state as `global` explicitly in each function. More importantly, having it all in a class allows the state to be quickly reset - this isn't currently used because reattach isn't fully implemented yet, but the intent is to throw away all those singletons on disconnect, and recreate them on reconnect.

Reply:
For this particular class, the third reason is as explained in the comment. With top-level functions, it'd require an import in each function, and it would also be slower because those imports would run (even if just to check `sys.modules` and bail if that module is already imported) every time the function executes.